platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+
+    for (Bundle bundle : bundleContext.getBundles()) {
+      if (toStop.contains(bundle.getSymbolicName())) {
+        try {
+          bundle.stop();
+        } catch (BundleException e) {
+          throw new SynchronisedInstallerException(
+              "Failed to stop bundle [" + bundle.getSymbolicName() + "]");
+        }
+      }
+    }
+
+    LOGGER.info("Finished stopping bundles in [{}] ms", (System.currentTimeMillis() - startTime));
+  }
+
+  @Override
+  public void startBundles(String symbolicName, String... additionalSymbolicNames)
+      throws SynchronisedInstallerException {
+    Set<String> toStart = toSet(symbolicName, additionalSymbolicNames);
+    LOGGER.info("Starting the following bundles:[{}]", String.join(", ", toStart));
+    long startTime = System.currentTimeMillis();
+
+    for (Bundle bundle : bundleContext.getBundles()) {
+      if (toStart.contains(bundle.getSymbolicName())) {
+        try {
+          bundle.start();
+        } catch (BundleException e) {
+          throw new SynchronisedInstallerException(
+              "Failed to start bundle [" + bundle.getSymbolicName() + "]");
+        }
+      }
+    }
+
+    LOGGER.info("Finished starting bundles in [{}] ms", (System.currentTimeMillis() - startTime));
+  }
+
+  @Override
+  public void waitForBundles(String... symbolicNames) throws SynchronisedInstallerException {
+    waitForBundles(DEFAULT_MAX_BUNDLE_WAIT, symbolicNames);
+  }
+
+  @Override
+  public void waitForBundles(long maxWaitTime, String... symbolicNames)
+      throws SynchronisedInstallerException {
+    Set<String> toWaitFor =
+        symbolicNames.length == 0
+            ? Arrays.stream(bundleContext.getBundles())
+                .map(Bundle::getSymbolicName)
+                .collect(Collectors.toSet())
+            : Arrays.stream(symbolicNames).collect(Collectors.toSet());
+
+    Callable<Boolean> areBundlesReady = () -> getUnavailableBundles(toWaitFor).isEmpty();
+
+    boolean available = wait(areBundlesReady, maxWaitTime, DEFAULT_POLLING_INTERVAL);
+    if (!available) {
+      printBundleInformation(LOGGER::error, getUnavailableBundles(toWaitFor));
+      throw new SynchronisedInstallerException(
+          "Bundles failed to reach Active within: " + maxWaitTime + " ms.");
+    }
+  }
+
+  @VisibleForTesting
+  protected boolean wait(
+      Callable<Boolean> conditionIsMet, long maxWait, long pollInterval, String onFailureMessage)
+      throws SynchronisedInstallerException {
+    if (pollInterval > maxWait) {
+      throw new SynchronisedInstallerException("Poll interval cannot be larger than maxWait time.");
+    }
+
+    final long startTime = System.currentTimeMillis();
+    LOGGER.trace("Waiting for condition to be met. Max wait time: [{}] ms", maxWait);
+
+    do {
+      try {
+        if (conditionIsMet.call()) {
+          return true;
+        } else {
+          Thread.sleep(pollInterval);
+        }
+      } catch (InterruptedException e) {
```

Comment:
+1 ... add a call to `Thread.interrupt()` here so the thread's interrupted state is not lost (flag is set behind the scenes).

Reply:
It is actually more than simply interrupt. You need to keep track of a boolean flag `interrupted` and set it to true and outside of the loop, put a try/finally where you check if that flag is true and re-interrupt the thread there. And normally, the first thing to do in the loop is calculated how much time left and only wait for that assuming it is greater than 0 or throw a timeout exception.

❗️ However, I am not sure it is a good idea here to delay interruption like this. Interruption will occur when a system is shutting down for example, in which case you don't want to be re-trying like this. For a utility library like this, the best course of action is to let the interruption bubble out by declaring it can be thrown by all methods that wait and let the callers handle that accordingly.